src/Compilers/Core/Portable/InternalUtilities/FatalError.cs

```diff
@@ -124,7 +132,7 @@ public static bool ReportAndPropagateUnlessCanceled(Exception exception, ErrorSe
         [DebuggerHidden]
         public static bool ReportAndPropagateUnlessCanceled(Exception exception, CancellationToken contextCancellationToken, ErrorSeverity severity = ErrorSeverity.Uncategorized)
         {
-            if (ExceptionUtilities.IsCurrentOperationBeingCancelled(exception, contextCancellationToken))
+            if (ExceptionUtilities.IsCurrentOperationBeingCancelled(exception, contextCancellationToken) || exception is OperationCanceledNotMachingCancellationTokenException)
```

Comment:
Why not just put this check in `IsCurrentOperationBeingCancelled`?

Also, instead of a custom exception type, would it be easier to have an overload of `IsCurrentOperationBeingCancelled` that doesn't take a cancellation token at all, so the caller can decide whether they care about the tokens matching, without having to alter other places to throw a custom exception?

Reply:
> Also, instead of a custom exception type, would it be easier to have an overload of IsCurrentOperationBeingCancelled that doesn't take a cancellation token at all, so the caller can decide whether they care about the tokens matching

Roslyn doesn't check for matching cancellation tokens.

Reply:
By matching I mean OperationCanceledException is thrown even when the cancellation token that flows into method is not signaled. Clearly Roslyn is checking that, this method does. If the call ends up calling back from OOP to devenv we would report NFE for that exception.